src/api/exchange/ExchangeApi.ts

```diff
@@ -97,21 +118,183 @@ export interface Order {
   remainingAmount: BN
 }
 
+/**
+ * BaseTradeEvent uses only info available on the emitted Trade event
+ */
+export interface BaseTradeEvent {
+  // order related
+  orderId: string
+  sellTokenId: number
+  buyTokenId: number
+  sellAmount: BN
+  buyAmount: BN
+  // block related
+  txHash: string
+  eventIndex: number
+  blockNumber: number
+  id: string // txHash | eventIndex
+}
+
+/**
+ * Trade enriches BaseTradeEvent with block, order and token data
+ */
+export interface Trade extends BaseTradeEvent {
+  batchId: number
+  revertKey: string // batchId | orderId, to find reverts
+  // indexOnBatch: number // tracks trade position on batch, in case of reverts
+  timestamp: number
+  settlingDate: Date
+  buyToken: TokenDetails
+  sellToken: TokenDetails
+  limitPrice: BigNumber
+  fillPrice: BigNumber
+  remainingAmount: BN
+}
+
 export interface GetOrdersPaginatedResult {
   orders: AuctionElement[]
   nextIndex?: number
 }
 
+// TODO: move to const/config
+const CONTRACT_DEPLOYMENT_BLOCK = {
+  1: 9340147,
+  4: 5844678,
+}
+
+type TradeEvent = BatchExchangeEvents['Trade']
+type TradeSubscription = Subscription<TradeEvent>
+
+interface Subscriptions {
+  trade: { [networkId: number]: TradeSubscription }
+}
 /**
  * Basic implementation of Stable Coin Converter API
  */
 export class ExchangeApiImpl extends DepositApiImpl implements ExchangeApi {
+  private subscriptions: Subscriptions = { trade: {} }
+
   public constructor(injectedDependencies: DepositApiDependencies) {
     super(injectedDependencies)
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
     ;(window as any).exchange = this._contractPrototype
   }
 
+  // TODO: Not very happy with this method. Can't be used inside the class because batchId is only known with block data
+  // TODO: Don't really know where to put it
+  public static buildTradeRevertKey(batchId: number, orderId: string): string {
+    return batchId + '|' + orderId
+  }
+
+  public async getPastTrades({
```

Comment:
You will eventually run into 
```
inpage.js:1 MetaMask - RPC Error: query returned more than 10000 results {code: -32005, message: "query returned more than 10000 results"}
```

Can get that even now with

```js
await exchangeApi.getPastTrades({networkId:1})
```

And as soon as any `userAddress` accumulates more than 10000 events, it'll break for that address

Reply:
Scheisse. 

I don't see a pagination API for this on web3. 

The best approach then is to fetch events in batches of arbitrary size and hope there are no more than 1k events per batch?

Reply:
you can do what we did somewhere before, @anxolin remembers
```js
const fetch = (...) => try {
 // fetch everything
} catch(error is our error) {
 fetch(from block to middle of prev range).concat(
   fetch(from middle of prev range)
  )
}
```
Binary fetch, like binary search

Reply:
70e7cdb82b8c342b369d818c35ab848d96c7d495

Reply:
> You will eventually run into...

Hehe, @Velenir you have good old memories from dxDAO times :)